core/src/main/scala/org/apache/spark/deploy/SparkS3Util.scala

```diff
@@ -0,0 +1,342 @@
+    for (obj: S3ObjectSummary <- objects.asScala) {
+      if (!obj.getKey.endsWith("/")) {
+        val path = new Path(bucket + "/" + obj.getKey)
+        if (hiddenFileFilter.accept(path)) {
+          val status: FileStatus = new FileStatus(
+            obj.getSize,
+            false,
+            1,
+            blockSize,
+            obj.getLastModified.getTime,
+            path)
+          list += status
+        }
+      }
+    }
+    list.iterator
+  }
+
+  /**
+   * For the given path, list S3 objects and return an iterator of returned FileStatuses.
+   */
+  @throws(classOf[AmazonClientException])
+  @throws(classOf[AmazonServiceException])
+  private def listPrefix(s3: AmazonS3Client, path: Path): Iterator[FileStatus] = {
+    val uri: URI = path.toUri
+    val key: String = keyFromPath(path)
+    val request: ListObjectsRequest = new ListObjectsRequest()
+      .withBucketName(uri.getAuthority)
+      .withPrefix(key)
+
+    val listings = new AbstractSequentialIterator[ObjectListing](s3.listObjects(request)) {
+      protected def computeNext(previous: ObjectListing): ObjectListing = {
+        if (!previous.isTruncated) {
+          return null
+        }
+        s3.listNextBatchOfObjects(previous)
+      }
+    }.asScala
+
+    val bucket: String = uri.getScheme + "://" + uri.getAuthority
+    listings
+      .map(listing => statusFromObjects(bucket, listing.getObjectSummaries))
+      .reduceLeft(_ ++ _)
+  }
+
+  /**
+   * For the given list of paths, sequentially list S3 objects per path and combine returned
+   * FileStatuses into a single array.
+   */
+  private def listStatus(s3: AmazonS3Client, paths: List[Path]): Array[FileStatus] = {
+    val list: ArrayBuffer[FileStatus] = ArrayBuffer()
+    paths.foreach { path =>
+      val iterator = listPrefix(s3, path)
+      while (iterator.hasNext) {
+        list += iterator.next
+      }
+    }
+    list.toArray
+  }
+
+  /**
+   * Find S3 block size from Hadoop conf. Try both s3 and s3n names.
+   */
+  private def getS3BlockSize(): Long = {
+    val minS3BlockSize = 10485760; // 10mb
+    val defaultS3BlockSize = 67108864; // 64mb
+    val value = Option(conf.get("fs.s3.block.size"))
+      .getOrElse(conf.get("fs.s3n.block.size", defaultS3BlockSize.toString)).toLong
+    if (value < minS3BlockSize) {
+      logWarning("S3 block size is set too small: " + value + ". Overriding it to 10mb.");
+      minS3BlockSize
+    } else {
+      value
+    }
+  }
+
+  /**
+   * Find min split size from Hadoop conf. Try both Hadoop 1 and 2 names.
+   */
+  private def getMinSplitSize(): Long = {
+    val value = Option(conf.get("mapred.min.split.size"))
+      .getOrElse(conf.get("mapreduce.input.fileinputformat.split.minsize", "134217728"))
+    value.toLong
+  }
+
+  /**
+   * Return whether the given path is an S3 path or not.
+   */
+  private def isS3Path(path: Path): Boolean = {
+    Option(path.toUri.getScheme).exists(_.toUpperCase.startsWith("S3"))
```

Comment:
I think you should avoid getting involved with S3N. In fact, if you really want to do it right, grab the schema, look up fs.${schema}.impl and look for it being one of `org.apache.hadoop.fs.s3.S3FileSystem` or
`org.apache.hadoop.fs.s3native.NativeS3FileSystem`, so if someone ever rebinds s3n to the s3a impl, things don't get confused.

If this code gets involved with s3a, then you may confuse some of the stuff there, especially HADOOP-11918 (listing empty dir -> FNFE) and in HADOOP-12537, its sts support (though that's not shipped yet). S3a uses the same bulk listing API and should therefore be something you can just leave alone.